include/deal.II/matrix_free/matrix_free.templates.h

```diff
@@ -59,6 +66,133 @@ MatrixFree<dim, Number>::MatrixFree(const MatrixFree<dim,Number> &other)
 
 
 
+template <int dim, typename Number>
+std::pair<unsigned int,unsigned int>
+MatrixFree<dim,Number>::create_cell_subrange_hp_by_index
+(const std::pair<unsigned int,unsigned int> &range,
+ const unsigned int fe_index,
+ const unsigned int vector_component) const
+{
+  AssertIndexRange (fe_index, dof_info[vector_component].max_fe_index);
+  const std::vector<unsigned int> &fe_indices =
+    dof_info[vector_component].cell_active_fe_index;
+  if (fe_indices.empty() == true)
+    return range;
+  else
+    {
+      // the range over which we are searching must be ordered, otherwise we
+      // got a range that spans over too many cells
+#ifdef DEBUG
+      for (unsigned int i=range.first+1; i<range.second; ++i)
+        Assert (fe_indices[i] >= fe_indices[i-1],
+                ExcMessage ("Cell range must be over sorted range of fe indices in hp case!"));
+      AssertIndexRange(range.first,fe_indices.size()+1);
+      AssertIndexRange(range.second,fe_indices.size()+1);
+#endif
+      std::pair<unsigned int,unsigned int> return_range;
+      return_range.first =
+        std::lower_bound(&fe_indices[0] + range.first,
+                         &fe_indices[0] + range.second, fe_index)
+        -&fe_indices[0] ;
+      return_range.second =
+        std::lower_bound(&fe_indices[0] + return_range.first,
+                         &fe_indices[0] + range.second,
+                         fe_index + 1)-&fe_indices[0];
+      Assert(return_range.first >= range.first &&
+             return_range.second <= range.second, ExcInternalError());
+      return return_range;
+    }
+}
+
+
+
+template <int dim, typename Number>
+void
+MatrixFree<dim,Number>::renumber_dofs (std::vector<types::global_dof_index> &renumbering,
+                                       const unsigned int vector_component)
+{
+  AssertIndexRange(vector_component, dof_info.size());
+  dof_info[vector_component].compute_dof_renumbering (renumbering);
+}
+
+
+
+template <int dim, typename Number>
+const DoFHandler<dim> &
+MatrixFree<dim,Number>::get_dof_handler (const unsigned int dof_index) const
+{
+  AssertIndexRange (dof_index, n_components());
+  if (dof_handlers.active_dof_handler == DoFHandlers::usual)
+    {
+      AssertDimension (dof_handlers.dof_handler.size(),
+                       dof_handlers.n_dof_handlers);
+      return *dof_handlers.dof_handler[dof_index];
+    }
+  else
+    {
+      Assert (false, ExcNotImplemented());
+      // put pseudo return argument to avoid compiler error, but trigger a
+      // segfault in case this is only run in optimized mode
```

Comment:
In that case why not using `AssertThrow` and get the stack trace directly without using `gdb`.